demos/dotnet/dotnet/ProjectPropertiesHelpers.cs

```diff
@@ -0,0 +1,335 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Xml;
+
+namespace dotnet
+{
+    internal static class ProjectPropertiesHelpers
+    {
+        public static ProjectProperties InitializeProperties(Settings settings, Log log)
+        {
+            // General Properites
+            var properties = new ProjectProperties();
+            var currentDirectory = Directory.GetCurrentDirectory();
+
+            var buildDll = settings.Target == "library";
+
+            properties.ProjectDirectory = Path.Combine(currentDirectory);
+            properties.PackagesDirectory = Path.Combine(properties.ProjectDirectory, "packages");
+            properties.OutputDirectory = Path.Combine(properties.ProjectDirectory, "bin");
+            properties.ToolsDirectory = Path.Combine(properties.ProjectDirectory, "tools");
+            properties.AssemblyName = Path.GetFileName(properties.ProjectDirectory);
+            properties.OutputType = buildDll ? ".dll" : ".exe";
+            FindCompiler(properties);
+
+            AddToListWithoutDuplicates(properties.Sources, ParseProjectFile(properties, settings.ProjectFile, "Compile"));
+            
+            foreach (var file in settings.SourceFiles.Where(f => !f.StartsWith(properties.PackagesDirectory) && !f.StartsWith(properties.OutputDirectory) &&
+                                                              !f.StartsWith(properties.ToolsDirectory)))
+            {
+                AddToListWithoutDuplicates(properties.Sources, file);
+            }
+
+            if (!string.IsNullOrWhiteSpace(settings.Recurse) && settings.Recurse != "*.cs")
+            {
+                AddToListWithoutDuplicates(properties.Sources,
+                    Directory.GetFiles(properties.ProjectDirectory, settings.Recurse, SearchOption.AllDirectories));
+            }
+
+            if (properties.Sources.Count == 1)
+            {
+                properties.AssemblyName = Path.GetFileNameWithoutExtension(properties.Sources[0]);
+            }
+
+            var platformOptionSpecicifcation = GetPlatformOption(settings.Platform);
+
+            // The anycpu32bitpreferred setting is valid only for executable (.EXE) files
```

Comment:
We should also build a list of strings, then add the preamble to that list of strings. It's more work to understand the code like this and there is more opportunity for error.